src/Turtle/Prelude.hs

```diff
@@ -1090,6 +1093,32 @@ time io = do
 hostname :: MonadIO io => io Text
 hostname = liftIO (fmap Text.pack getHostName)
 
+-- | Show the full path of an executable file
+which :: MonadIO io => FilePath -> io (Maybe FilePath)
+which cmd = fold go Control.Foldl.head
+  where
+    go :: Shell FilePath
+    go =
+      need "PATH" >>= \case
+          Nothing -> empty
+          Just paths -> do
+            path <- select (Text.split (== ':') paths)
+            let path' = Filesystem.fromText path </> cmd
+            exists <- testfile path'
+            if exists
+                then do
+                    perms <- liftIO $
+                        getmod path'
+                          `catchIOError`
+                            \e ->
```

Comment:
One minor suggestion to improve readability.  Perhaps factor out the handler as a named function, like this:

```haskell
...
then do
    let handler e =
            if isPermissionError e || isDoesNotExistError e
                then return Directory.emptyPermissions
                else throwIO e
    perms <- liftIO (getmod path' `catchIOError` handler)
    ...
```

The other reason that I like this trick is that you can add a type signature to `handler` to specify the type of the exception to catch (although it's not necessary in this example)